core/src/impl/Kokkos_ViewArray.hpp

```diff
@@ -390,22 +390,33 @@ class ViewMapping<Traits, Kokkos::Array<>> {
                                     alloc_size)
             : record_type::allocate(mem_space, alloc_name, alloc_size);
 
-    if (alloc_size) {
-      m_impl_handle =
-          handle_type(reinterpret_cast<pointer_type>(record->data()));
-
-      if (alloc_prop::initialize) {
-        // The functor constructs and destroys
-        record->m_destroy =
-            execution_space_specified
-                ? functor_type(exec_space, (pointer_type)m_impl_handle,
-                               m_impl_offset.span() * Array_N, alloc_name)
-                : functor_type((pointer_type)m_impl_handle,
-                               m_impl_offset.span() * Array_N, alloc_name);
+    m_impl_handle = handle_type(reinterpret_cast<pointer_type>(record->data()));
 
+    functor_type functor =
+        execution_space_specified
+            ? functor_type(exec_space, (pointer_type)m_impl_handle,
+                           m_impl_offset.span() * Array_N, alloc_name)
+            : functor_type((pointer_type)m_impl_handle,
+                           m_impl_offset.span() * Array_N, alloc_name);
+
+    if (false) {
+      // Make sure the destroy functor gets instantiated.
+      // This avoids "cudaErrorInvalidDeviceFunction"-type errors.
+      functor.destroy_shared_allocation();
+    }
+
+    //  Only initialize if the allocation is non-zero.
+    //  May be zero if one of the dimensions is zero.
+    if constexpr (alloc_prop::initialize)
+      if (alloc_size) {
+        // Assume destruction is only required when construction is requested.
+        // The ViewValueFunctor has both value construction and destruction
+        // operators.
+        record->m_destroy = std::move(functor);
```

Comment:
I see that is what we were currently doing but does that mean we don't setup for destructing the elements when a view was constructed specifying the "without initializing"?
Is that a potential issue?

Reply:
Yes, we only call the destructor for non-`std::is_scalar`-types and on those only if the `View` constructor also initializes the elements. This design dates back all the way to 95bc2e1657a1f77d725721b1774cb7715ab22907 but it doesn't really give a justification for this assumption. I think it's a question to be discussed separately from this pull request.

Reply:
We discussed on Slack that we want to keep the current behavior.